File: 2024/day13b.c
// 18446744072854832140 : trop grand

#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <math.h>
#include <assert.h>

#define COUT_A 3
#define COUT_B 1
#define A_AJOUTER 10000000000000

typedef long long int llint;
typedef unsigned long long int ullint;

bool divise(ullint num, ullint denom) {
    return denom * (num / denom) == num;
}

ullint min_tokens(ullint xA, ullint yA, ullint xB, ullint yB, ullint px, ullint py) {
    ullint det = (xA * yB) - (xB * yA);
    ullint numA, numB;
    if (det < 0) {
        numA = - (yB * px) + (xB * py);
        numB = - (xA * py) + (yA * px);
    }
    else {
        numA = (yB * px) - (xB * py);
        numB = (xA * py) - (yA * px);
    }
    printf("   Pour %llu %llu : %llu %llu\n", px, py, numA, numB);
    if (divise(numA, det) && divise(numB, det)) {
        return (COUT_A * numA / det) + (COUT_B * numB / det);
    }
    return 1;
}

int lire_machine_suivante(FILE* f) {
    ullint xA, yA, xB, yB, px, py;
    if (fscanf(f, "Button A: X+%llu, Y+%llu\n", &xA, &yA) != 2) {
        return 2;
    }
    fscanf(f, "Button B: X+%llu, Y+%llu\n", &xB, &yB);
    fscanf(f, "Prize: X=%llu, Y=%llu\n", &px, &py);
    fscanf(f, "\n");
    return min_tokens(xA, yA, xB, yB, px+A_AJOUTER, py+A_AJOUTER);
}

int main() {
    FILE* f = fopen("day13.txt", "r");
    ullint total = 0;
    ullint val = 0;
    while (val != 2) {
        if (val != 1) total += val;
        val = lire_machine_suivante(f);
        printf("%llu\n", val);
    }
    fclose(f);
    printf("TOTAL : %llu\n", total);
    return 0;
}

File: 2024/day13b.py
COUT_A = 3
COUT_B = 1

total = 0

def divise(num, denom):
    return denom * (num // denom) == num

with open("day13_extraction.txt") as fichier:
    for i in range(320):
        boutonA = fichier.readline().split(" ")
        boutonB = fichier.readline().split(" ")
        pos_prix = fichier.readline().split(" ")
        fichier.readline()
        print(boutonA, boutonB, pos_prix)
        xA = int(boutonA[0])
        yA = int(boutonA[1])
        xB = int(boutonB[0])
        yB = int(boutonB[1])
        px = int(pos_prix[0]) + 10000000000000
        py = int(pos_prix[1]) + 10000000000000
        det = (xA * yB) - (xB * yA)
        numA = (yB * px) - (xB * py)
        numB = (xA * py) - (yA * px)
        if divise(numA, det) and divise(numB, det):
            total += (COUT_A * numA // det) + (COUT_B * numB // det)
print(total)